File: aspnet/src/ShowAd.AspNetCore/Jwt/JwtHelper.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace ShowAd.AspNetCore.Jwt;

/// <summary>
/// Minimal JWT decoder. The backend handles signature verification; this
/// helper only inspects payload claims (exp, nbf, creator_hash, fingerprint, iss).
///
/// Defense-in-depth: tokens whose header `alg` is `none` or outside the
/// HS256/HS384/HS512/RS256/RS384/RS512/ES256/ES384 whitelist are rejected
/// before any payload claims are inspected.
/// </summary>
public static class JwtHelper
{
    public const string ExpectedIssuer = "showad-backend";

    public const long DefaultLeewaySeconds = 60;

    /// <summary>Algorithms accepted for local payload inspection.</summary>
    public static readonly HashSet<string> AllowedAlgorithms = new(StringComparer.Ordinal)
    {
        "HS256", "HS384", "HS512",
        "RS256", "RS384", "RS512",
        "ES256", "ES384",
    };

    public readonly record struct ValidationResult(bool Valid, string? Reason);

    public sealed class ClaimValidationOptions
    {
        public long LeewaySeconds { get; init; } = DefaultLeewaySeconds;
        public bool RequireIssuer { get; init; } = true;
    }

    public static IReadOnlyDictionary<string, JsonElement>? DecodeToken(string? token)
    {
        if (string.IsNullOrWhiteSpace(token)) return null;
        var parts = token.Split('.');
        if (parts.Length != 3) return null;

        // Reject 'none' and unknown algorithms (defense in depth).
        try
        {
            var headerBytes = Base64UrlDecode(parts[0]);
            using var headerDoc = JsonDocument.Parse(headerBytes);
            if (headerDoc.RootElement.ValueKind != JsonValueKind.Object) return null;
            if (!headerDoc.RootElement.TryGetProperty("alg", out var alg)
                || alg.ValueKind != JsonValueKind.String
                || alg.GetString() is not { } algStr
                || !AllowedAlgorithms.Contains(algStr))
            {
                return null;
            }
        }
        catch
        {
            return null;
        }

        byte[] payload;
        try { payload = Base64UrlDecode(parts[1]); }
        catch { return null; }

        try
        {
            using var doc = JsonDocument.Parse(payload);
            if (doc.RootElement.ValueKind != JsonValueKind.Object) return null;
            var dict = new Dictionary<string, JsonElement>(StringComparer.Ordinal);
            foreach (var p in doc.RootElement.EnumerateObject())
            {
                dict[p.Name] = p.Value.Clone();
            }
            return dict;
        }
        catch (JsonException)
        {
            return null;
        }
    }

    public static bool IsTokenExpired(string? token, DateTimeOffset? now = null, long leewaySeconds = DefaultLeewaySeconds)
    {
        var claims = DecodeToken(token);
        if (claims is null) return true;

        var nowSec = (now ?? DateTimeOffset.UtcNow).ToUnixTimeSeconds();

        if (claims.TryGetValue("exp", out var exp) && TryGetLong(exp, out var expVal) && (expVal + leewaySeconds) < nowSec)
            return true;

        if (claims.TryGetValue("nbf", out var nbf) && TryGetLong(nbf, out var nbfVal) && (nbfVal - leewaySeconds) > nowSec)
            return true;

        if (claims.TryGetValue("iat", out var iat) && TryGetLong(iat, out var iatVal) && (iatVal - leewaySeconds) > nowSec)
            return true;

        return false;
    }

    /// <summary>Returns the token expiry as Unix seconds (matches JWT `exp` claim), or null when missing.</summary>
    public static long? GetTokenExpiry(string? token)
    {
        var claims = DecodeToken(token);
        if (claims is null) return null;
        if (!claims.TryGetValue("exp", out var exp)) return null;
        return TryGetLong(exp, out var v) ? v : null;
    }

    public static ValidationResult ValidateTokenClaims(
        string? token,
        string expectedCreatorHash,
        string? expectedFingerprint = null,
        DateTimeOffset? now = null,
        ClaimValidationOptions? options = null)
    {
        options ??= new ClaimValidationOptions();
        var claims = DecodeToken(token);
        if (claims is null)
            return new ValidationResult(false, "Invalid token format");

        if (IsTokenExpired(token, now, options.LeewaySeconds))
            return new ValidationResult(false, "Token expired");

        if (!claims.TryGetValue("creator_hash", out var ch) || ch.ValueKind != JsonValueKind.String
            || !FixedTimeEqual(ch.GetString(), expectedCreatorHash))
        {
            return new ValidationResult(false, "Creator hash mismatch");
        }

        if (expectedFingerprint is not null)
        {
            if (!claims.TryGetValue("fingerprint", out var fp) || fp.ValueKind != JsonValueKind.String
                || !FixedTimeEqual(fp.GetString(), expectedFingerprint))
            {
                return new ValidationResult(false, "Fingerprint mismatch");
            }
        }

        var issuerProvided = claims.TryGetValue("iss", out var iss) && iss.ValueKind == JsonValueKind.String;
        if (options.RequireIssuer)
        {
            if (!issuerProvided || !string.Equals(iss.GetString(), ExpectedIssuer, StringComparison.Ordinal))
            {
                return new ValidationResult(false, "Invalid issuer");
            }
        }
        else if (issuerProvided && !string.Equals(iss.GetString(), ExpectedIssuer, StringComparison.Ordinal))
        {
            return new ValidationResult(false, "Invalid issuer");
        }

        return new ValidationResult(true, null);
    }

    public static string? GetCreatorHashFromToken(string? token)
    {
        var claims = DecodeToken(token);
        if (claims is null) return null;
        return claims.TryGetValue("creator_hash", out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;
    }

    public static string? GetFingerprintFromToken(string? token)
    {
        var claims = DecodeToken(token);
        if (claims is null) return null;
        return claims.TryGetValue("fingerprint", out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;
    }

    private static bool FixedTimeEqual(string? a, string? b)
    {
        if (a is null || b is null) return false;
        var aBytes = Encoding.UTF8.GetBytes(a);
        var bBytes = Encoding.UTF8.GetBytes(b);
        if (aBytes.Length != bBytes.Length) return false;
        return CryptographicOperations.FixedTimeEquals(aBytes, bBytes);
    }

    private static bool TryGetLong(JsonElement el, out long value)
    {
        switch (el.ValueKind)
        {
            case JsonValueKind.Number:
                if (el.TryGetInt64(out value)) return true;
                if (el.TryGetDouble(out var d)) { value = (long)d; return true; }
                value = 0; return false;
            case JsonValueKind.String:
                return long.TryParse(el.GetString(), out value);
            default:
                value = 0; return false;
        }
    }

    public static byte[] Base64UrlDecode(string value)
    {
        var s = value.Replace('-', '+').Replace('_', '/');
        switch (s.Length % 4)
        {
            case 2: s += "=="; break;
            case 3: s += "="; break;
            case 1: throw new FormatException("Invalid base64url length");
        }
        return Convert.FromBase64String(s);
    }

    public static string Base64UrlEncode(byte[] data)
    {
        return Convert.ToBase64String(data).TrimEnd('=').Replace('+', '-').Replace('/', '_');
    }

    public static string Base64UrlEncode(string data) => Base64UrlEncode(Encoding.UTF8.GetBytes(data));
}

File: php/src/Jwt/JwtHelper.php
<?php

declare(strict_types=1);

namespace ProofMark\ShowAd\Jwt;

/**
 * JWT decoding helper.
 *
 * IMPORTANT: this helper does NOT verify the signature. The ShowAd backend
 * issues and verifies tokens; the SDK only inspects unsigned claims to make
 * cheap, local decisions (expiry, creator hash match, fingerprint match).
 * Treat decoded values as untrusted hints, not authoritative facts.
 *
 * Defense-in-depth: rejects tokens whose header `alg` is `none` or outside the
 * HS256/HS384/HS512/RS256/RS384/RS512/ES256/ES384 whitelist.
 */
final class JwtHelper
{
    public const ISSUER = 'showad-backend';

    /** Tokens signed with these algorithms are accepted for local inspection. */
    public const ALLOWED_ALGORITHMS = [
        'HS256', 'HS384', 'HS512',
        'RS256', 'RS384', 'RS512',
        'ES256', 'ES384',
    ];

    public const DEFAULT_LEEWAY_SECONDS = 60;

    /**
     * Decode the payload section of a JWT into an associative array.
     *
     * @return array<string, mixed>|null
     */
    public static function decodeToken(string $token): ?array
    {
        $parts = explode('.', $token);
        if (count($parts) !== 3) {
            return null;
        }

        $headerJson = self::base64UrlDecode($parts[0]);
        if ($headerJson === null) {
            return null;
        }
        $header = json_decode($headerJson, true);
        if (!is_array($header) || !isset($header['alg']) || !is_string($header['alg'])) {
            return null;
        }
        if (!in_array($header['alg'], self::ALLOWED_ALGORITHMS, true)) {
            return null;
        }

        $payload = self::base64UrlDecode($parts[1]);
        if ($payload === null) {
            return null;
        }

        $claims = json_decode($payload, true);
        if (!is_array($claims)) {
            return null;
        }

        return $claims;
    }

    public static function isTokenExpired(string $token, ?int $now = null, int $leewaySeconds = self::DEFAULT_LEEWAY_SECONDS): bool
    {
        $claims = self::decodeToken($token);
        if ($claims === null) {
            return true;
        }

        $now = $now ?? time();

        if (isset($claims['exp']) && is_numeric($claims['exp']) && ((int) $claims['exp'] + $leewaySeconds) < $now) {
            return true;
        }

        if (isset($claims['nbf']) && is_numeric($claims['nbf']) && ((int) $claims['nbf'] - $leewaySeconds) > $now) {
            return true;
        }

        if (isset($claims['iat']) && is_numeric($claims['iat']) && ((int) $claims['iat'] - $leewaySeconds) > $now) {
            return true;
        }

        return false;
    }

    /**
     * Returns the token expiry as Unix seconds (matches JWT `exp` claim).
     */
    public static function getTokenExpiry(string $token): ?int
    {
        $claims = self::decodeToken($token);
        if ($claims === null || !isset($claims['exp']) || !is_numeric($claims['exp'])) {
            return null;
        }

        return (int) $claims['exp'];
    }

    public static function getTimeUntilExpiry(string $token, ?int $now = null): int
    {
        $expiry = self::getTokenExpiry($token);
        if ($expiry === null) {
            return -1;
        }

        $now = $now ?? time();
        return $expiry - $now;
    }

    /**
     * Validate the locally-checkable claims of a token.
     *
     * @param array{leeway_seconds?: int, require_issuer?: bool} $options
     * @return array{valid: bool, reason: ?string}
     */
    public static function validateTokenClaims(
        string $token,
        string $expectedCreatorHash,
        ?string $expectedFingerprint = null,
        array $options = []
    ): array {
        $leeway = isset($options['leeway_seconds']) ? (int) $options['leeway_seconds'] : self::DEFAULT_LEEWAY_SECONDS;
        $requireIssuer = array_key_exists('require_issuer', $options) ? (bool) $options['require_issuer'] : true;

        $claims = self::decodeToken($token);
        if ($claims === null) {
            return ['valid' => false, 'reason' => 'Invalid token format'];
        }

        if (self::isTokenExpired($token, null, $leeway)) {
            return ['valid' => false, 'reason' => 'Token expired'];
        }

        if (!isset($claims['creator_hash']) || !is_string($claims['creator_hash'])
            || !hash_equals((string) $claims['creator_hash'], $expectedCreatorHash)) {
            return ['valid' => false, 'reason' => 'Creator hash mismatch'];
        }

        if ($expectedFingerprint !== null) {
            if (!isset($claims['fingerprint']) || !is_string($claims['fingerprint'])
                || !hash_equals((string) $claims['fingerprint'], $expectedFingerprint)) {
                return ['valid' => false, 'reason' => 'Fingerprint mismatch'];
            }
        }

        if ($requireIssuer) {
            if (!isset($claims['iss']) || $claims['iss'] !== self::ISSUER) {
                return ['valid' => false, 'reason' => 'Invalid issuer'];
            }
        } elseif (isset($claims['iss']) && $claims['iss'] !== self::ISSUER) {
            return ['valid' => false, 'reason' => 'Invalid issuer'];
        }

        return ['valid' => true, 'reason' => null];
    }

    public static function getCreatorHashFromToken(string $token): ?string
    {
        $claims = self::decodeToken($token);
        if ($claims === null || !isset($claims['creator_hash'])) {
            return null;
        }
        return (string) $claims['creator_hash'];
    }

    public static function getFingerprintFromToken(string $token): ?string
    {
        $claims = self::decodeToken($token);
        if ($claims === null || !isset($claims['fingerprint'])) {
            return null;
        }
        return (string) $claims['fingerprint'];
    }

    private static function base64UrlDecode(string $value): ?string
    {
        $value = strtr($value, '-_', '+/');
        $padding = strlen($value) % 4;
        if ($padding > 0) {
            $value .= str_repeat('=', 4 - $padding);
        }
        $decoded = base64_decode($value, true);
        return $decoded === false ? null : $decoded;
    }
}
